cmd/xl-v1-healing.go

```diff
@@ -82,29 +86,56 @@ func (xl xlObjects) HealBucket(bucket string) error {
 	}); reducedErr != nil {
 		return toObjectErr(reducedErr, bucket)
 	}
-	return nil
+
+	// Proceed to heal bucket metadata.
+	return healBucketMetadata(xl.storageDisks, bucket)
 }
 
-// HealObject heals a given object for all its missing entries.
-// FIXME: If an object object was deleted and one disk was down, and later the disk comes back
-// up again, heal on the object should delete it.
-func (xl xlObjects) HealObject(bucket, object string) error {
-	// Verify if bucket is valid.
-	if !IsValidBucketName(bucket) {
-		return traceError(BucketNameInvalid{Bucket: bucket})
+// Heals all the metadata associated for a given bucket.
+func healBucketMetadata(storageDisks []StorageAPI, bucket string) error {
+	policyPath := pathJoin(bucketConfigPrefix, bucket, policyJSON)
+	// Lock the `policy.json` before healing.
+	policyLock := nsMutex.NewNSLock(minioMetaBucket, policyPath)
+	policyLock.RLock()
+	defer policyLock.RUnlock()
+
```

Comment:
> Since this mutates the file on on some disks, shouldn't this take a write lock instead of read?

The reason to not hold Write lock is that it would block the Read operations, also since we are only writing to missing disks a Write() on this will still block because of our Read lock. So all in all it works without much trouble. 

This is how Object healing is implemented and it is a correct behavior.